File: app/javascript/runtime/loading/resource_manifest.js
export const RESOURCE_CLASSES = Object.freeze(["critical", "contextual", "interaction", "viewport", "predictive", "idle"])

function strings(value, field) {
  if (!Array.isArray(value) || value.some((entry) => typeof entry !== "string" || !entry)) {
    throw new TypeError(`${field} must contain non-empty strings`)
  }
  return Object.freeze([...new Set(value)])
}

export class ResourceManifest {
  constructor(input = {}) {
    if (typeof input.context !== "string" || !input.context.match(/^[a-z0-9_.-]+$/i)) {
      throw new TypeError("manifest context is invalid")
    }

    this.context = input.context
    this.styles = strings(input.styles || [], "styles")
    this.controllers = strings(input.controllers || [], "controllers")
    this.media = Object.freeze({ ...(input.media || {}) })
    this.audio = Object.freeze({ unlock: false, cues: [], bed: null, ...(input.audio || {}) })
    this.motion = strings(input.motion || [], "motion")
    this.prefetch = Object.freeze({ nextScreen: false, maxBytes: 0, ...(input.prefetch || {}) })
    this.classes = Object.freeze({ ...(input.classes || {}) })

    if (!Array.isArray(this.audio.cues) || this.audio.cues.some((cue) => typeof cue !== "string")) {
      throw new TypeError("audio cues must be strings")
    }
    if (this.prefetch.maxBytes < 0 || this.prefetch.maxBytes > 180_000) {
      throw new RangeError("prefetch maxBytes exceeds the runtime budget")
    }
    Object.values(this.classes).forEach((kind) => {
      if (!RESOURCE_CLASSES.includes(kind)) throw new TypeError(`unknown resource class: ${kind}`)
    })
    Object.freeze(this)
  }

  static fromElement(element) {
    if (!element?.textContent) return new ResourceManifest({ context: "shell", styles: ["shell"] })
    return new ResourceManifest(JSON.parse(element.textContent))
  }
}

File: app/services/frontend/resource_manifest.rb
module Frontend
  class ResourceManifest
    CLASSES = %w[critical contextual interaction viewport predictive idle].freeze
    MAX_PREFETCH_BYTES = 180_000

    attr_reader :context, :styles, :controllers, :media, :audio, :motion, :prefetch, :classes

    def self.shell
      new(
        context: "shell",
        styles: %w[shell],
        controllers: %w[loading press],
        classes: {
          "style.shell" => "critical",
          "controller.loading" => "critical",
          "controller.press" => "contextual"
        }
      )
    end

    def initialize(context:, styles: [], controllers: [], media: {}, audio: {}, motion: [], prefetch: {}, classes: {})
      @context = context.to_s
      @styles = normalize_list(styles, :styles)
      @controllers = normalize_list(controllers, :controllers)
      @media = media.to_h.compact.deep_stringify_keys.freeze
      @audio = { "unlock" => false, "cues" => [], "bed" => nil }.merge(audio.to_h.deep_stringify_keys).freeze
      @motion = normalize_list(motion, :motion)
      @prefetch = { "nextScreen" => false, "maxBytes" => 0 }.merge(prefetch.to_h.deep_stringify_keys).freeze
      @classes = classes.to_h.transform_keys(&:to_s).transform_values(&:to_s).freeze
      validate!
      freeze
    end

    def as_json(*)
      {
        context:,
        styles:,
        controllers:,
        media:,
        audio:,
        motion:,
        prefetch:,
        classes:
      }
    end

    def to_json(*)
      as_json.to_json
    end

    private

      def normalize_list(value, name)
        raise ArgumentError, "#{name} must be an array" unless value.is_a?(Array)

        value.map(&:to_s).reject(&:blank?).uniq.freeze
      end

      def validate!
        raise ArgumentError, "invalid frontend context" unless context.match?(/\A[a-z0-9_.-]+\z/i)
        raise ArgumentError, "audio cues must be an array" unless audio["cues"].is_a?(Array)

        bytes = Integer(prefetch["maxBytes"])
        raise ArgumentError, "prefetch budget exceeds #{MAX_PREFETCH_BYTES}" unless bytes.between?(0, MAX_PREFETCH_BYTES)

        unknown = classes.values - CLASSES
        raise ArgumentError, "unknown resource classes: #{unknown.join(', ')}" if unknown.any?
      end
  end
end
